lib/extractor/layer.ts

```diff
@@ -0,0 +1,141 @@
+    });
+
+    tarExtractor.on("finish", () => {
+      resolve(getLayersContent(layersNames, layers));
+    });
+
+    tarExtractor.on("error", reject);
+
+    createReadStream(dockerArchiveFilesystemPath).pipe(tarExtractor);
+  });
+}
+
+/**
+ * Extract key files from the specified TAR stream.
+ * @param layerTarStream image layer as a Readable TAR stream. Note: consumes the stream.
+ * @param extractActions array of pattern, callbacks pairs
+ * @returns extracted file products
+ */
+export async function extractImageLayer(
+  layerTarStream: Readable,
+  extractActions: ExtractAction[],
+): Promise<ExtractedLayers> {
+  return new Promise((resolve, reject) => {
+    const result: ExtractedLayers = {};
+    const tarExtractor: Extract = extract();
+
+    tarExtractor.on("entry", async (headers, stream, next) => {
+      if (headers.type === "file") {
+        const absoluteFileName = `/${headers.name}`;
+        const processedResult = await extractFileAndProcess(
+          absoluteFileName,
+          stream,
+          extractActions,
+        );
+        if (processedResult !== undefined) {
+          result[absoluteFileName] = processedResult;
+        }
+      }
+
+      stream.resume(); // auto drain the stream
+      next(); // ready for next entry
+    });
+
+    tarExtractor.on("finish", () => {
+      // all layer level entries read
+      resolve(result);
+    });
+
+    tarExtractor.on("error", reject);
+
+    layerTarStream.pipe(tarExtractor);
+  });
+}
+
+/**
+ * Note: consumes the stream.
+ */
+async function extractFileAndProcess(
+  fileName: string,
+  fileStream: Readable,
+  extractActions: ExtractAction[],
+): Promise<FileNameAndContent | undefined> {
+  const matchedActions = extractActions.filter((action) =>
+    minimatch(fileName, action.fileNamePattern, { dot: true }),
+  );
+
+  if (matchedActions.length > 0) {
+    return await applyCallbacks(matchedActions, fileStream);
+  }
+
+  return undefined;
+}
+
+function getLayersContent(
+  layersNames: string[],
+  layers: Record<string, ExtractedLayers>,
+): ExtractedLayers[] {
+  // skip (ignore) non-existent layers
+  // get the layers content without the name
+  // reverse layers order from last to first
+  return layersNames
+    .filter((layersName) => layers[layersName])
+    .map((layerName) => layers[layerName])
+    .reverse();
+}
+
+/**
+ * Note: consumes the stream.
+ */
+function getLayersNames(stream: Readable): Promise<string[]> {
+  return streamToString(stream).then((manifestFile) => {
+    const manifest = JSON.parse(manifestFile);
+    return manifest[0].Layers;
+  });
+}
+
+function isManifestFile(name: string) {
+  return name === "manifest.json";
+}
+
+function isTarFile(name: string) {
+  // For both "docker save" and "skopeo copy" style archives the
+  // layers are represented as tar archives whose names end in .tar.
+  // For Docker this is "layer.tar", for Skopeo - "<sha256ofLayer>.tar".
+  return basename(name).indexOf(".tar") !== -1;
```

Comment:
Consider the following change
```suggestion
  return basename(name).endsWith(".tar");
```